pkg/operator/status.go

```diff
@@ -66,182 +91,225 @@ func (r *StatusReporter) GetOrCreateClusterOperator() (*configv1.ClusterOperator
-	// cluster autoscalers are at the latest version.
-	AvailableAndUpdated() (bool, error)
+	return r.ApplyStatus(status)
 }
 
 // progressing reports the operator as progressing but available, and not
 // failing -- optionally setting a reason and message.
-func (r *StatusReporter) progressing() error {
-	glog.Infof("Syncing to version %v", r.releaseVersion)
-	conditions := []configv1.ClusterOperatorStatusCondition{
-		{
-			Type:   configv1.OperatorAvailable,
-			Status: configv1.ConditionTrue,
-		},
-		{
-			Type:    configv1.OperatorProgressing,
-			Status:  configv1.ConditionTrue,
-			Reason:  ReasonSyncing,
-			Message: fmt.Sprintf("Syncing to version %v", r.releaseVersion),
-		},
-		{
-			Type:   configv1.OperatorFailing,
-			Status: configv1.ConditionFalse,
+func (r *StatusReporter) progressing(reason, message string) error {
+	status := configv1.ClusterOperatorStatus{
+		Conditions: []configv1.ClusterOperatorStatusCondition{
+			{
+				Type:   configv1.OperatorAvailable,
+				Status: configv1.ConditionTrue,
+			},
+			{
+				Type:    configv1.OperatorProgressing,
+				Status:  configv1.ConditionTrue,
+				Reason:  reason,
+				Message: message,
+			},
+			{
+				Type:   configv1.OperatorFailing,
+				Status: configv1.ConditionFalse,
+			},
 		},
 	}
 
-	return r.ApplyConditions(conditions, false)
+	glog.Infof("Operator status progressing: %s", message)
+
+	return r.ApplyStatus(status)
 }
 
-// applyStatusInterface required here for test coverage of critical code.
-// In actual operation, we are passing in StatusReporter.
-func applyStatus(r applyStatusInterface, check AvailableChecker) (bool, error) {
-	ok, err := r.checkMachineAPI()
-	if err != nil {
-		r.fail(ReasonMissingDependency, fmt.Sprintf("error checking machine-api operator status %v", err))
-		return false, nil
-	}
-	if !ok {
-		r.fail(ReasonMissingDependency, "machine-api operator not ready")
-		return false, nil
+// Start checks the status of dependencies and reports the operator's status. It
+// will poll until stopCh is closed or prerequisites are satisfied, in which
+// case it will report the operator as available the configured version and wait
+// for stopCh to close before returning.
+func (r *StatusReporter) Start(stopCh <-chan struct{}) error {
+	interval := 15 * time.Second
+
+	// Poll the status of our prerequisites and set our status
+	// accordingly.  Rather than return errors and stop polling, most
+	// errors here should just be reported in the status message.
+	pollFunc := func() (bool, error) {
+		return r.ReportStatus()
 	}
-	ok, err = check.AvailableAndUpdated()
+
+	err := wait.PollImmediateUntil(interval, pollFunc, stopCh)
+
+	// Block until the stop channel is closed.
+	<-stopCh
+
+	return err
+}
+
+// ReportStatus checks the status of each dependency and operand and reports the
+// appropriate status via the operator's ClusterOperator object.
+func (r *StatusReporter) ReportStatus() (bool, error) {
```

Comment:
This is a good place for passing in StatusReporter as an interface instead of using a method.  This would make mocking up the logic of this function easier, and it would decouple the implementation from the spec.

Reply:
I get what you're saying, but I just don't think that necessary here. This is already easy to test by just mocking up the state of the things it checks like we do in the rest of the tests. I think that's a more realistic test as opposed to passing in a mock implementation.